crates/ruff_linter/src/rules/airflow/rules/removal_in_3.rs

```diff
@@ -1,12 +1,130 @@
 use ruff_diagnostics::{Diagnostic, Edit, Fix, FixAvailability, Violation};
 use ruff_macros::{derive_message_formats, ViolationMetadata};
-use ruff_python_ast::{name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall};
+use ruff_python_ast::{
+    name::QualifiedName, Arguments, Expr, ExprAttribute, ExprCall, ExprContext, ExprName,
+    StmtClassDef,
+};
 use ruff_python_semantic::analyze::typing;
 use ruff_python_semantic::Modules;
+use ruff_python_semantic::ScopeKind;
 use ruff_text_size::Ranged;
 
 use crate::checkers::ast::Checker;
 
+fn is_airflow_secret_backend(segments: &[&str]) -> bool {
```

Comment:
Maybe we could combine all the functions into a single function and accept the changing variables? I think those are the module names and the ends_with value. What about the following?

```rust
fn is_airflow_builtin_or_provider(segments: &[&str], module: &str, symbol_suffix: &str) -> bool {
    match segments {
        ["airflow", "secrets", rest @ ..] => {
            if let Some(last_element) = rest.last() {
                last_element.ends_with(symbol_suffix)
            } else {
                false
            }
        }

        ["airflow", "providers", rest @ ..] => {
            if let (Some(pos), Some(last_element)) =
                (rest.iter().position(|&s| s == module), rest.last())
            {
                pos + 1 < rest.len() && last_element.ends_with(symbol_suffix)
            } else {
                false
            }
        }

        _ => false,
    }
}
```

Reply:
I tried to refactor the utility functions a bit. Please take a look when you're available. Thanks!